File: src/cliente/ClienteChat/controlador/ChatControlador.cs
using System;
using System.Reflection.Metadata;
using System.Text.Json.Nodes;
using System.Text.Json;
using ClienteChat.modelo;
using ClienteChat.servicios;
using ClienteChat.vista;
using Microsoft.VisualBasic;

// Aquí vamos a coordinar todo nuestro MVC para el cliente
namespace ClienteChat.controlador
{
    public class ChatControlador
    {
        private NetworkService _networkService;
        private VistaConsola _view;
        private UsuarioModelo _usuario;

        // Aquí creamos el controlador
        public ChatControlador(NetworkService networkService, VistaConsola view)
        {
            _networkService = networkService;
            _view = view;
            _usuario = new UsuarioModelo();

            // Y usamos los eventos de la capa de red
            _networkService.OnMessageReceived += ProcesarMensajeServidor;
            _networkService.OnDisconnected += ManejarDesconexion;
        }

        // Ahora, haremos el método (que ya teníamos) para identificar el tipo de comando que escribió el usuario, solo que lo vamos a modificar
        public void IdentificarTipo(String entrada)
        {
            // Vamos a partir la entrada para identificar el comando que introdujo el usuario y todo del argumento que dió
            string[] partes = entrada.Split(' ', 2);
            string comando = partes[0].ToLower(); // Aquí guardamos el comando
            string argumento; // Aquí guardamos el argumento del usuario en caso de que lo haya
            if(partes.Length > 1)
            {
                argumento = partes[1];
            } else
            {
                argumento = string.Empty;
            }

            // Aquí haremos una verificación para no dejar al usuario hacer nada en caso de que no se haya identificado todavía
            if(comando != "/identify" && !_usuario.EstaIdentificado)
            {
                _view.MostrarError("Debes identificarte primero usando el comando: /identify <username>");
                return;
            }

            // Usaremos un switch para simplificar todos los ifs que tenía
            switch(comando)
            {
                // El comando es IDENTIFY
                case "/identify":
                    Identify(argumento);
                    break;
                
                // El comando es STATUS
                case "/status":
                    Status(argumento);
                    break;
                
                // El comando es USERS
                case "/users":
                    Users();
                    break;
                
                // El comando es TEXT
                case "/text":
                    Text(argumento);
                    break;
                
                // El comando es PUBLIC_TEXT
                case "/public_text":
                    PublicText(argumento);
                    break;
                
                // El comando es NEW_ROOM
                case "/new_room":
                    NewRoom(argumento);
                    break;
                
                // El comando es INVITE
                case "/invite":
                    Invite(argumento);
                    break;

                // El comando es JOIN_ROOM
                case "/join_room":
                    JoinRoom(argumento);
                    break;

                // El comando es ROOM_USERS
                case "/room_users":
                    RoomUsers(argumento);
                    break;

                default:
                _view.MostrarError($"Introduzca un comando válido ({comando} no lo es).");
                break;
            }
        }

        // Aquí abajo ya van todos los métodos para comunicar al cliente con el servidor mediante mensajes JSON

        // Definición de
        // IDENTIFY (del lado del cliente)
        // :)
        public void Identify(String username)
        {
            // Verificamos que username no sea null o un espacio en blanco
            if(string.IsNullOrWhiteSpace(username))
            {
                _view.MostrarError("Uso correcto: /identify <username>");
                return;
            }

            // Verificamos que el nombre de usuario no sea mayor a 8 caracteres
            if(username.Length <= 8)
            {
                // Empezamos a armar nuestro JSON
                var iden_json = new JsonObject();

                iden_json["type"] = "IDENTIFY"; // Decimos que el tipo de comando es IDENTIFY
                iden_json["username"] = username; // Decimos que el username es el especificado por el usuario
                string json_texto = iden_json.ToJsonString(); // Convertimos el JSON en un string

                _networkService.Enviar(json_texto);
                _usuario.Username = username;
                    
            } else
            {
                _view.MostrarError("El nombre debe ser de menos de 8 caracteres.");
            }
        }

        // Definición de
        // STATUS (del lado del cliente)
        // :)
        public void Status(String entrada)
        {
            string estado = entrada.ToUpper();

            // Comprobamos que su estado sea válido
            if(estado.Equals("ACTIVE") || estado.Equals("AWAY") || estado.Equals("BUSY"))
            {
                // Ahora sí empezamos a armar nuestro JSON
                var stat_json = new JsonObject();

                stat_json["type"] = "STATUS"; // Decimos que el tipo de comando es STATUS
                stat_json["status"] = estado; // Decimos que el status es el especificado por el usuario (pero en mayúsculas)
                string json_texto = stat_json.ToJsonString(); // Convertimos el JSON a un string

                _networkService.Enviar(json_texto); // Enviamos el texto
                _usuario.Status = estado;
            } else
            {
                _view.MostrarError("El estado que quieres debe ser uno de los tres válidos: ACTIVE, AWAY o BUSY."); // Informamos al usuario que la regó
            }
        }

        // Definición de
        // USERS (del lado del cliente)
        // :)
        public void Users()
        {
            // Como aquí es simplemente un comando y no hay nada más de información importante, no hace falta ni que usemos la entrada del usuario
            var users_json = new JsonObject();

            users_json["type"] = "USERS";

            // Ps un poco más de lo de siempre, mandamos el JSON al servidor y ya :v
            string json_texto = users_json.ToJsonString();
            _networkService.Enviar(json_texto);

            // No puedo creer lo sencilla que fue esta parte
        }

        // Definición de
        // TEXT (del lado del cliente)
        // :)
        public void Text(String entrada)
        {
            // Vamos a partir la entrada porque el argumento debería contener al username del destinatario y el texto que se le quiere mandar
            string[] partes = entrada.Split(' ', 2);

            // Revisamos que no falte información importante
            if(partes.Length < 2)
            {
                _view.MostrarError("Falta información en el comando. Uso correcto: /text <username> <texto>");
                return;
            }

            // Empezamos a crear nuestro JSON
            var text_json = new JsonObject();

            text_json["type"] = "TEXT"; // Decimos que es de tipo TEXT
            text_json["username"] = partes[0]; // Decimos quién es el destinatario
            text_json["text"] = partes[1]; // Decimos que el mensaje es el resto de la entrada

            // Por como cuarta vez, mandamos el JSON :v
            string json_texto = text_json.ToJsonString();
            _networkService.Enviar(json_texto);
        }

        // Definición de
        // PUBLIC_TEXT (del lado del cliente)
        // :)
        public void PublicText(String mensaje)
        {
            // Verificamos que el mensaje no sea null o un espacio en blanco
            if(string.IsNullOrWhiteSpace(mensaje))
            {
                _view.MostrarError("Uso correcto: /public_text <mensaje>");
                return;
            }

            // Empezamos a crear el JSON que vamos a mandarle al servidor
            var publictext_json = new JsonObject();

            publictext_json["type"] = "PUBLIC_TEXT"; // Decimos que es de tipo PUBLIC_TEXT
            publictext_json["text"] = mensaje; // Decimos que el mensaje es el resto de la entrada

            // Por quinta vez y contandoooo mandamos el JSON
            string json_texto = publictext_json.ToJsonString();
            _networkService.Enviar(json_texto);
        }

        // Definición de
        // NEW_ROOM (del lado del cliente)
        // :)
        public void NewRoom(String roomname)
        {
            // Primero verifiquemos que el nombre no sea NULL o un espacio en blanco
            if(string.IsNullOrWhiteSpace(roomname))
            {
                _view.MostrarError("Debes proporcionar un nombre para el cuarto. Uso correcto: /new_room <nombre_del_cuarto>");
                return;
            }
            // Verificamos que el nombre sea de a lo mucho 16 caracteres
            if(roomname.Length <= 16)
            {
                // Ya que haya pasado esas dos verificaciones, ahora sí podemos empezar a trabajar
                var newroom_json = new JsonObject();

                newroom_json["type"] = "NEW_ROOM";
                newroom_json["roomname"] = roomname;

                // Por sexta vez siiiii mandamos el JSON
                string json_texto = newroom_json.ToJsonString();
                _networkService.Enviar(json_texto);
            } else
            {
                _view.MostrarError("El nombre de cuarto solo puede ser de máximo 16 caracteres");
            }
        }

        // Definición de
        // INVITE (del lado del cliente)
        // :)
        public void Invite(String entrada)
        {
            // Primero verificamos que no falte información importante
            if(string.IsNullOrWhiteSpace(entrada))
            {
                _view.MostrarError("Falta información. Uso correcto (la sala debe estar entre comillas): /invite <sala> <usuarios_a_invitar>");
                return;
            }

            // Primero vamos a partir la entrada usando comillas porque el nombre de la sala debe estar entre comillas
            string[] partes = entrada.Split('"');

            if(partes.Length < 3)
            {
                _view.MostrarError("Formato de comillas incorrecto. Uso correcto (la sala debe estar entre comillas): /invite <sala> <usuarios_a_invitar>");
                return;
            }

            string room_name = partes[1].Trim();
            string[] usuariosAEnviar = partes[2].Split(' ', StringSplitOptions.RemoveEmptyEntries);

            // Vamos a verificar rápido que por lo menos haya un usuario
            if(usuariosAEnviar.Length == 0)
            {
                _view.MostrarError("Debes proporcionar usuarios para invitar. Uso correcto (la sala debe estar entre comillas): /invite <sala> <usuarios_a_invitar>");
                return;
            }

            // Como ya verificamos que sí hay usuarios para invitar, hacemos un arreglo para almacenarlos
            var usernames = new JsonArray();

            for(int i = 0; i < usuariosAEnviar.Length; i++)
            {
                usernames.Add(usuariosAEnviar[i]);
            }

            // Por último, hacemos el JSON para mandarlo al servidor
            var invite_json = new JsonObject();

            invite_json["type"] = "INVITE";
            invite_json["roomname"] = room_name;
            invite_json["usernames"] = usernames;

            // Van siete veces que convertimos un json a string xd
            string json_texto = invite_json.ToJsonString();
            _networkService.Enviar(json_texto);
        }

        // Definición de
        // JOIN_ROOM
        // :)
        public void JoinRoom(String roomname)
        {
            // De nuevo verificamos que no falte información importante
            if(string.IsNullOrWhiteSpace(roomname))
            {
                _view.MostrarError("Falta información. Uso correcto: /join_room <sala>");
                return;
            }

            // La ventaja aquí es que, como la entrada es el nombre de la sala, no hay que partir nada
            var join_json = new JsonObject();

            join_json["type"] = "JOIN_ROOM";
            join_json["roomname"] = roomname;

            // Ocho veces mandando jsons siiiii
            string json_texto = join_json.ToJsonString();
            _networkService.Enviar(json_texto);
        }

        // Definición de
        // ROOM_USERS (del lado del cliente)
        // :)
        public void RoomUsers(String roomname)
        {
            // Otra vez verificamos que no falte información importante
            if(string.IsNullOrWhiteSpace(roomname))
            {
                _view.MostrarError("Falta información. Uso correcto: /room_users <sala>");
                return;
            }

            // Igual que con join_room, como solo es el nombre de la sala no hay que partir nada :D
            var room_users_json = new JsonObject();

            room_users_json["type"] = "ROOM_USERS";
            room_users_json["roomname"] = roomname;

            // VAN NUEVE veces que mandamos el json :D
            string json_texto = room_users_json.ToJsonString();
            _networkService.Enviar(json_texto);
        }

        // Ya estos métodos de aquí abajo sirve para escuchar las respuestas del servidor

        // Este método de aquí va a procesar los mensajes del servidor
        private void ProcesarMensajeServidor(string jsonRespuesta)
        {
            try
            {
                // Parseamos el JSON y obtenemos su raíz (pues los JSON son árboles si no me equivoco)
                var mensaje = JsonDocument.Parse(jsonRespuesta);
                var raiz = mensaje.RootElement;

                // Hacemos una pequeña comprobación por seguridad, en el mero caso de que no haya un "type" en el JSON
                if(raiz.TryGetProperty("type", out var type))
                {
                    string tipo = type.GetString() ?? "";

                    // En caso de que sea una respuesta
                    if(tipo == "RESPONSE")
                    {
                        // Guardamos la operación que se hizo y el resultado de la misma
                        string operation = raiz.GetProperty("operation").GetString() ?? "";
                        string result = raiz.GetProperty("result").GetString() ?? "";

                        if(result == "SUCCESS")
                        {
                            _view.MostrarInformacion($"Operación {operation} realizada exitosamente");
                            if(operation == "IDENTIFY") // Esta verificación de aquí es para decir que el usuario ya se identificó
                            {
                                _usuario.EstaIdentificado = true;
                            }
                        } else
                        {
                            _view.MostrarError($"Error en {operation}: {result}");
                        }
                    }
                    else if(tipo == "NEW_USER") // En caso de que sea una notificación de un nuevo usuario
                    {
                        string user = raiz.GetProperty("username").GetString() ?? "Anónimo";
                        _view.MostrarInformacion($"[Notificación]: {user} se ha conectado. ¡Dí hola!");
                    }
                    else if(tipo == "NEW_STATUS") // En caso de que sea una notificación de un nuevo estado de un usuario
                    {
                        string user = raiz.GetProperty("username").GetString() ?? "Anónimo";
                        string status = raiz.GetProperty("status").GetString() ?? "";
                        _view.MostrarInformacion($"[Notificación]: {user} ahora está {status}");
                    } 
                    else if(tipo == "USER_LIST") // En caso de que se haya solicitado la lista de usuarios
                    {
                        _view.MostrarInformacion("--- Lista de Usuarios Conectados ---");
                        if (raiz.TryGetProperty("users", out var users))
                        {
                            foreach (var userProp in users.EnumerateObject())
                            {
                                _view.MostrarMensaje($" • {userProp.Name} -> [{userProp.Value.GetString()}]");
                            }
                        }
                    }
                    else if(tipo == "TEXT_FROM") // En caso de que sea un mensaje privado
                    {
                        string emisor = raiz.GetProperty("username").GetString() ?? "Anónimo";
                        string message = raiz.GetProperty("text").GetString() ?? "";
                        _view.MostrarMensaje($"[Privado de: {emisor}]: {message}");
                    }
                    else if(tipo == "PUBLIC_TEXT_FROM"){
                        string emisor = raiz.GetProperty("username").GetString() ?? "Anónimo";
                        string message = raiz.GetProperty("text").GetString() ?? "";
                        _view.MostrarMensaje($"[Público de: {emisor}]: {message}");
                    }
                    else if(tipo == "INVITATION") // En caso de que sea una invitación
                    {
                        string emisor = raiz.GetProperty("username").GetString() ?? "Anónimo";
                        string roomname = raiz.GetProperty("roomname").GetString() ?? "";
                        _view.MostrarInformacion($"[Notificación]: ¡{emisor} te está invitando a la sala {roomname}!");
                    }
                    else if(tipo == "JOINED_ROOM") // En caso de que se haya aceptado la invitación
                    {
                        string user = raiz.GetProperty("username").GetString() ?? "Anónimo";
                        string roomname = raiz.GetProperty("roomname").GetString() ?? "";
                        _view.MostrarInformacion($"[Notificación]: ¡{user} acaba de entrar a {roomname}!");
                    }
                    else if(tipo == "ROOM_USER_LIST") // En caso de que se haya solicitado la lista de usuarios de una sala
                    {
                        string roomname = raiz.GetProperty("roomname").GetString() ?? "";

                        _view.MostrarInformacion($"--- Lista de Usuarios Conectados en {roomname} ---");
                        if (raiz.TryGetProperty("users", out var users))
                        {
                            foreach (var userProp in users.EnumerateObject())
                            {
                                _view.MostrarMensaje($" • {userProp.Name} -> [{userProp.Value.GetString()}]");
                            }
                        }
                    }
                    else
                    {
                        // Ya muestra cualquier cosa que haya mandado el servidor
                        _view.MostrarInformacion($"[Servidor]: {jsonRespuesta}");
                    }
                }
            } catch
            {
                // Voy a dejar que no haga nada en caso de que lo que se manda no sea un json
            }
        }

        // Este método de aquí es para manejar la desconexión de parte del servidor
        private void ManejarDesconexion()
        {
            _view.MostrarError("Se perdió la conexión con el servidor.");
        }
    }

}

File: src/servidor/controlador/chat_controlador.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "chat_controlador.h"
#include "../vista/red_vista.h"

// Este método lo usamos para identificar el tipo que viene en el JSON
void identificar_tipo(cJSON *json, int socket_cliente, ListaClientes *lista_clientes, ListaSalas *lista_salas) {
    cJSON *type = cJSON_GetObjectItemCaseSensitive(json, "type"); // En type vamos a guardar el tipo de mensaje que mandó el cliente

    // Ahora aquí vamos a revisar cuál es el tipo para saber a qué otra función debe llamar esta función
    if(cJSON_IsString(type) && (type->valuestring != NULL)) {
        // Revisamos a qué función pertenece el tipo

        // En caso de que type sea IDENTIFY
        if(strcmp(type->valuestring, "IDENTIFY") == 0) {
            identify(lista_clientes, socket_cliente, json);
        }

        // En caso de que type sea STATUS
        if(strcmp(type->valuestring, "STATUS") == 0) {
            status(socket_cliente, json, lista_clientes);
        }

        // En caso de que type sea USERS
        if(strcmp(type->valuestring, "USERS") == 0) {
            users(socket_cliente, lista_clientes);
        }

        // En caso de que sea type TEXT
        if(strcmp(type->valuestring, "TEXT") == 0) {
            text(socket_cliente, json, lista_clientes);
        }

        // En caso de que sea type PUBLIC_TEXT
        if(strcmp(type->valuestring, "PUBLIC_TEXT") == 0) {
            publictext(socket_cliente, json, lista_clientes);
        }

        // En caso de que sea type NEW_ROOM
        if(strcmp(type->valuestring, "NEW_ROOM") == 0) {
            newroom(socket_cliente, json, lista_salas, lista_clientes);
        }

        // En caso de que sea type INVITE
        if(strcmp(type->valuestring, "INVITE") == 0) {
            invite(socket_cliente, json, lista_salas, lista_clientes);
        }
    }
}

// Aquí ya van el resto de métodos individuales

// Definición de:
// IDENTIFY
// :)
void identify(ListaClientes *lista, int socket_cliente, cJSON *json) {
    cJSON *username = cJSON_GetObjectItemCaseSensitive(json, "username");
    cJSON *response = cJSON_CreateObject();

    cJSON_AddStringToObject(response, "type", "RESPONSE");
    cJSON_AddStringToObject(response, "operation", "IDENTIFY");

    // Aquí revisamos que el username no exista ya
    if(buscar_cliente(lista, username->valuestring) == 0) {
        insertar_cliente(lista, socket_cliente, username->valuestring, "ACTIVE");

        // Empezamos a formar todo el JSON de respuesta
        cJSON_AddStringToObject(response, "result", "SUCCESS");
        cJSON_AddStringToObject(response, "extra", username->valuestring);

        // Ahora hacemos el JSON para todos los demás usuarios
        cJSON *notif = cJSON_CreateObject();
        cJSON_AddStringToObject(notif, "type", "NEW_USER");
        cJSON_AddStringToObject(notif, "username", username->valuestring);

        // Ahora mandamos el mensaje a todos los demás usuarios
        notificar_usuarios_vista(lista, socket_cliente, notif);
        
        // Luego liberamos memoria
        cJSON_Delete(notif);
    } else {
        // Aquí formamos el JSON de respuesta en caso de que ya exista el username
        cJSON_AddStringToObject(response, "result", "USER_ALREADY_EXISTS");
        cJSON_AddStringToObject(response, "extra", username->valuestring);
    }

    enviar_json(socket_cliente, response);

    // Ahora liberamos toooda la memoria usada
    cJSON_Delete(response);
}

// Definición de
// STATUS
// :)
void status(int socket_cliente, cJSON *json, ListaClientes *lista) {
    cJSON *status = cJSON_GetObjectItemCaseSensitive(json, "status");

    char *nuevo_status = status->valuestring;

    // Esta parte de aquí la usaremos para poder actualizar el estado del cliente en la lista
    ClienteNodo *actual = lista->cabeza;
    
    while(actual != NULL) {
        if(actual->socket == socket_cliente) {
            // Esta comprobación también la hacemos en el lado del cliente así que no impora mucho xd, lo único que importa aquí es ver si el estado del usuario es igual al que quiere poner
            if((strcmp(nuevo_status, "ACTIVE") == 0 || strcmp(nuevo_status, "AWAY") == 0 || strcmp(nuevo_status, "BUSY") == 0) && strcmp(nuevo_status, actual->status) != 0) {
                // En caso de que pase la comprobación, actualizamos el estado del usuario
                strncpy(actual->status, status->valuestring, sizeof(actual->status) - 1);
                actual->status[sizeof(actual->status) - 1] = '\0';

                // Aquí formaremos el JSON y se lo mandaremos a los demás usuarios

                cJSON *notif = cJSON_CreateObject();
                cJSON_AddStringToObject(notif, "type", "NEW_STATUS");
                cJSON_AddStringToObject(notif, "username", actual->username);
                cJSON_AddStringToObject(notif, "status", nuevo_status);

                notificar_usuarios_vista(lista, socket_cliente, notif);

                // Liberamos la memoria utilizada
                cJSON_Delete(notif);
                return;
            }
        }

        actual = actual->siguiente;
    }
}

// Definición de
// USERS
// :)
void users(int socket_cliente, ListaClientes *lista) {
    // Empezaremos directamente creando la respuesta pues es la parte más simple del protocolo (hasta ahora)
    cJSON *response = cJSON_CreateObject();
    cJSON_AddStringToObject(response, "type", "USER_LIST");

    // Ahora vamos a crear un JSON que almacenará toooooda la lista de usuarios
    cJSON *user_list = cJSON_CreateObject();

    // Y ahora vamos a recorrer toda la lista de usuarios para poder añadirlos a la lista
    ClienteNodo *actual = lista->cabeza;
    while(actual != NULL) {
        cJSON_AddStringToObject(user_list, actual->username, actual->status);
        actual = actual->siguiente;
    }

    // Metemos la lista de usuarios al JSON
    cJSON_AddItemToObject(response, "users", user_list);

    // Lo mandamos al usuario que lo solicitó
    enviar_json(socket_cliente, response);

    cJSON_Delete(response);
}

// Definición de
// TEXT
// :)
void text(int socket_cliente, cJSON *json, ListaClientes *lista) {
    cJSON *username = cJSON_GetObjectItemCaseSensitive(json, "username");
    cJSON *message = cJSON_GetObjectItemCaseSensitive(json, "text");

    // Primero, confirmaremos que en efecto existe el usuario al que se le quiere mandar el mensaje
    if(buscar_cliente(lista, username->valuestring) == 1) {

        //Vamos a iterar en la lista de usuarios, por lo que vamos a guardar el socket del usuario al que se le quiere mandar el mensaje
        int socket_destino = 0;
        // También como soy tonto, voy a usar el while que ya teníamos para poder obtener el username de la persona que manda el mensaje xd
        char *username_emisor;

        ClienteNodo *actual = lista->cabeza;

        while(actual != NULL) {
            // Si esta condición de aquí se cumple, quiere decir que ya encontró al usuario al que se le quiere mandar mensaje
            if(strcmp(actual->username, username->valuestring) == 0) {
                socket_destino = actual->socket;
            }

            // Esta mausquerramienta de aquí nos ayudará más tarde (encontrar el username del emisor)
            if(actual->socket == socket_cliente) {
                username_emisor = actual->username;
            }

            // Si no pasa la condición, revisamos si el siguiente en la lista es el que buscamos
            actual = actual->siguiente;
        }

        // Ya que haya salido del while, entonces vamos a formar el json que contendrá al mensaje
        cJSON *notif = cJSON_CreateObject();
        cJSON_AddStringToObject(notif, "type", "TEXT_FROM");
        cJSON_AddStringToObject(notif, "username", username_emisor);
        cJSON_AddStringToObject(notif, "text", message->valuestring);

        // Y ps lo mandamos
        enviar_json(socket_destino, notif);

        cJSON_Delete(notif);

    } else { // En caso de que no exista, el servidor responderá
        cJSON *response = cJSON_CreateObject();
        cJSON_AddStringToObject(response, "type", "RESPONSE");
        cJSON_AddStringToObject(response, "operation", "TEXT");
        cJSON_AddStringToObject(response, "result", "NO_SUCH_USER");
        cJSON_AddStringToObject(response, "extra", username->valuestring);

        enviar_json(socket_cliente, response);

        cJSON_Delete(response);
    }
}

// Definición de
// PUBLIC_TEXT
// :)
void publictext(int socket_cliente, cJSON *json, ListaClientes *lista) {
    cJSON *message = cJSON_GetObjectItemCaseSensitive(json, "text");

    // En este caso, no hay que revisar si el usuario existe porque pues, es para todos, si acaso lo que hay que revisar es si hay más de un usuario, para que en caso de que no, no haga nada
    if(lista->numClientes > 1) {
        // Hay que iterar en la lista para poder saber quién mandó el mensaje xd
        char *username_emisor;

        ClienteNodo *actual = lista->cabeza;
        while(actual != NULL) {
            // Ha vuelto la mausquerramienta
            if(actual->socket == socket_cliente) {
                username_emisor = actual->username;
            }
            actual = actual->siguiente;
        }

        // Empezamos a formar el JSON
        cJSON *notif = cJSON_CreateObject();
        cJSON_AddStringToObject(notif, "type", "PUBLIC_TEXT_FROM");
        cJSON_AddStringToObject(notif, "username", username_emisor);
        cJSON_AddStringToObject(notif, "text", message->valuestring);

        // Y se lo mandamos a todoooos (menos al usuario que lo mandó obvio xd)
        notificar_usuarios_vista(lista, socket_cliente, notif);

        cJSON_Delete(notif);
    }
}

// Definición de
// NEW_ROOM
// :)
void newroom(int socket_cliente, cJSON *json, ListaSalas *lista_salas, ListaClientes *lista) {
    cJSON *room_name = cJSON_GetObjectItemCaseSensitive(json, "roomname");
    cJSON *response = cJSON_CreateObject();

    // Primero vamos a comprobar que la sala no exista ya, en caso de que sí pues le decimos al usuario que ya existe una sala con ese nombre
    if(buscar_sala(lista_salas, room_name->valuestring) == 1) {
        cJSON_AddStringToObject(response, "type", "RESPONSE");
        cJSON_AddStringToObject(response, "operation", "NEW_ROOM");
        cJSON_AddStringToObject(response, "result", "ROOM_ALREADY_EXISTS");
        cJSON_AddStringToObject(response, "extra", room_name->valuestring);
    } else {
        // En caso de que no exista ya, la creamos :D
        insertar_sala(lista_salas, room_name->valuestring);

        // Ahora vamos a buscar el username y el estado del usuario que creó la sala
        ClienteNodo *actual = lista->cabeza;
        char *username;
        char *status;
        while(actual != NULL) {
            if(actual->socket == socket_cliente) {
                username = actual->username;
                status = actual->status;
            }

            actual = actual->siguiente;
        }

        // Ahora vamos a meter al usuario que creó la sala
        NodoSala *nueva_sala = obtener_sala(lista_salas, room_name->valuestring);

        insertar_cliente(nueva_sala->usuarios, socket_cliente, username, status);

        // Por último, vamos a hacer el JSON de respuesta
        cJSON_AddStringToObject(response, "type", "RESPONSE");
        cJSON_AddStringToObject(response, "operation", "NEW_ROOM");
        cJSON_AddStringToObject(response, "result", "SUCCESS");
        cJSON_AddStringToObject(response, "extra", room_name->valuestring);
    }

    // Y lo mandamos
    enviar_json(socket_cliente, response);

    cJSON_Delete(response);
}

// Definición de
// INVITE
// :)
void invite(int socket_cliente, cJSON *json, ListaSalas *lista_salas, ListaClientes *lista) {
    cJSON *roomname = cJSON_GetObjectItemCaseSensitive(json, "roomname");
    cJSON *usernames = cJSON_GetObjectItemCaseSensitive(json, "usernames");

    if(!cJSON_IsString(roomname) || !cJSON_IsArray(usernames)) {
        return;
    }

    // Primero vamos a revisar que efectivamente existe la sala
    if(buscar_sala(lista_salas, roomname->valuestring) == 0) {
        // Si no existe, le notificaremos al usuario que puso una sala inexistente xd
        cJSON *response = cJSON_CreateObject();
        cJSON_AddStringToObject(response, "type", "RESPONSE");
        cJSON_AddStringToObject(response, "operation", "INVITE");
        cJSON_AddStringToObject(response, "result", "NO_SUCH_ROOM");
        cJSON_AddStringToObject(response, "extra", roomname->valuestring);

        enviar_json(socket_cliente, response);

        cJSON_Delete(response);
        return;
    }

    // Ahora, vamos a verificar que todos los usuarios especificados en el arreglo sí existen
    cJSON *user_item = NULL;
    cJSON_ArrayForEach(user_item, usernames) {
        if(cJSON_IsString(user_item) && user_item->valuestring != NULL) {
            if(buscar_cliente(lista, user_item->valuestring) == 0) {
                // Al detectar al primer usuario inexistente, respondemos que no existe
                cJSON *response = cJSON_CreateObject();
                cJSON_AddStringToObject(response, "type", "RESPONSE");
                cJSON_AddStringToObject(response, "operation", "INVITE");
                cJSON_AddStringToObject(response, "result", "NO_SUCH_USER");
                cJSON_AddStringToObject(response, "extra", user_item->valuestring);

                enviar_json(socket_cliente, response);

                cJSON_Delete(response);
                return;
            }
        }
    }

    // Ahora que ya sabemos que tanto la sala como los usuarios existen, vamos a obtener el nombre de usuario del que mandó la invitación
    ClienteNodo *actual = lista->cabeza;
    char *emisor;
    while(actual != NULL) {
        if(actual->socket == socket_cliente) {
            emisor = actual->username;
        }
        actual = actual->siguiente;
    }

    // Y por último vamos a mandar todas las invitaciones
    NodoSala *sala = obtener_sala(lista_salas, roomname->valuestring);

    cJSON_ArrayForEach(user_item, usernames) {
        // Si el usuario ya está en la sala simplemente lo vamos a ignorar
        if(sala != NULL && buscar_cliente(sala->usuarios, user_item->valuestring) == 1) {
            continue;
        }

        // Vamos ahora a buscar el socket del usuario al que se quiere invitar
        int socket_destino = -1;
        ClienteNodo *actual_destino = lista->cabeza;
        while(actual_destino != NULL) {
            if(strcmp(user_item->valuestring, actual_destino->username) == 0) {
                socket_destino = actual_destino->socket;
                break;
            }

            actual_destino = actual_destino->siguiente;
        }

        // Y por último vamos a mandarle la invitación al usuario
        if(socket_destino != -1) {
            cJSON *notif = cJSON_CreateObject();
            cJSON_AddStringToObject(notif, "type", "INVITATION");
            cJSON_AddStringToObject(notif, "username", emisor);
            cJSON_AddStringToObject(notif, "roomname", sala);

            enviar_json(socket_destino, notif);
            cJSON_Delete(notif);
        }
    }
}